### ISSUE TITLE
🐛 [SDAR] Potential Issue with BGPO Algorithm Implementation

### ISSUE THREAD

**Issue by yyloong** [yjyddq/DARE#18]
I noticed a potential issue in models/SDAR-8B-Chat/modeling_sdar.py when computing the old policy probability. It seems that the code enters the evaluation (eval) case, in which the block causal mask and random noise mask are not constructed. This might lead to incorrect importance ratio calculations. Could the maintainers please check whether the old policy probability computation should apply the same masking logic as in the training case, or if the current behavior is intentional? Thank you for your excellent work on the DARE project!

<img width="1189" height="754" alt="Image" src="https://github.com/user-attachments/assets/cacdfed4-b4af-4b64-8bdb-11b0dfebe528" />

**Comment by yjyddq** [comment]
Thanks for your attention. We will follow up on this issue.

**Comment by yjyddq** [comment]
Thanks again for reporting this issue!                                                                                                              
                                                                                                                                                    
We've pushed a fix in commit 4513b59. Here's what was changed and why:                                                                                                                                                                                                
### Root Cause                                                      
During BGPO training with the SDAR model, the old log-prob recomputation's `forward_process` is independently with log-prob ones. They go through different branches in `modeling_sdar.py`. This caused  a mismatch of `<MASK>` noise, leading to incorrect importance sampling.             
                                                                  
### What Was Fixed                                                                                                                                    
- `dllm_ray_trainer.py`: Removed the `if model.name != "sdar"` guard that was skipping forward_process for SDAR. SDAR now goes through `forward_process` like other models (e.g., LLaDA), so perturbed_seq, mask_indices, and p_mask are pre-computed and stored in the batch.
- `modeling_sdar.py` (all 3 variants): Added a new `prepare_for_bd_training_from_artifacts` method and a `force_block_diffusion_training` flag. The model reuses them together with noise in `forward_process` instead of re-sampling, ensuring consistency between log-prob and old log-prob.
- `sdar_dp_actor_bgpo.py`: Updated `_forward_micro_batch` and `_get_logits` to extract the pre-computed artifacts from the batch and pass them to the model.

**Comment by yjyddq** [comment]
Kindly re-open this issue if you still have problems.
